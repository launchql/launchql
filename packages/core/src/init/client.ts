import { Logger } from '@launchql/logger';
import { readFileSync } from 'fs';
import { join } from 'path';
import { Pool } from 'pg';
import { getPgPool } from 'pg-cache';
import { PgConfig } from 'pg-env';

const log = new Logger('init');

export class LaunchQLInit {
  private pool: Pool;
  private pgConfig: PgConfig;

  constructor(config: PgConfig) {
    this.pgConfig = config;
    this.pool = getPgPool(this.pgConfig);
  }

  /**
   * Bootstrap standard roles (anonymous, authenticated, administrator)
   */
  async bootstrapRoles(): Promise<void> {
    try {
      log.info('Bootstrapping LaunchQL roles...');
      
      const sqlPath = join(__dirname, 'sql', 'bootstrap-roles.sql');
      const sql = readFileSync(sqlPath, 'utf-8');
      
      await this.pool.query(sql);
      
      log.success('Successfully bootstrapped LaunchQL roles');
    } catch (error) {
      log.error('Failed to bootstrap roles:', error);
      throw error;
    }
  }

  /**
   * Bootstrap test roles (roles only, no users)
   */
  async bootstrapTestRoles(): Promise<void> {
    try {
      log.info('Bootstrapping LaunchQL test roles...');
      
      const sqlPath = join(__dirname, 'sql', 'bootstrap-test-roles.sql');
      const sql = readFileSync(sqlPath, 'utf-8');
      
      await this.pool.query(sql);
      
      log.success('Successfully bootstrapped LaunchQL test roles');
    } catch (error) {
      log.error('Failed to bootstrap test roles:', error);
      throw error;
    }
  }

  /**
   * Close the database connection
   */
  async close(): Promise<void> {
  }
}
