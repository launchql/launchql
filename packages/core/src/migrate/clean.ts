import { RawStmt } from '@pgsql/types';
import { parse, deparse } from 'pgsql-parser';

const filterStatements = (stmts: RawStmt[]): { filteredStmts: RawStmt[], hasFiltered: boolean } => {
  const filteredStmts = stmts.filter(node => {
    const stmt = node.stmt;
    return stmt && !stmt.hasOwnProperty('TransactionStmt') && 
           !stmt.hasOwnProperty('CreateExtensionStmt');
  });
  
  const hasFiltered = filteredStmts.length !== stmts.length;
  return { filteredStmts, hasFiltered };
};

export const cleanSql = async (
  sql: string, 
  pretty: boolean, 
  functionDelimiter: string
): Promise<string> => {
  const parsed = await parse(sql);
  const { filteredStmts, hasFiltered } = filterStatements(parsed.stmts as any);
  
  if (!hasFiltered) {
    return sql;
  }
  
  parsed.stmts = filteredStmts;
  
  const finalSql = await deparse(parsed, {
    pretty,
    functionDelimiter
  });

  return finalSql;
};
