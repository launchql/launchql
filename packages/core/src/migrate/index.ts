export * from './client';
export * from './types';
export * from './migration';
export * from '../modules/deploy';
export * from '../modules/revert';
export * from '../modules/verify';
export * from './clean';
export * from './utils/hash';
export * from './utils/transaction';
