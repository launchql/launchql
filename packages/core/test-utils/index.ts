export { TestPlan } from './TestPlan';
export { CoreDeployTestFixture } from './CoreDeployTestFixture';
export { TestFixture } from './TestFixture';

export { MigrateTestFixture } from './MigrateTestFixture';
export { TestDatabase } from './TestDatabase';
export { MigrateTestChange } from './MigrateTestChange';

export {
  FIXTURES_PATH,
  getFixturePath,
  cleanText,
  teardownAllPools,
  closeDatabasePools
} from './utils';
