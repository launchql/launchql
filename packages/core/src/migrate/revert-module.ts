import { join } from 'path';
import { existsSync } from 'fs';
import { LaunchQLMigrate } from '@launchql/migrate';
import { MigrateConfig } from '@launchql/migrate';
import { Logger } from '@launchql/logger';

const log = new Logger('migrate-revert');

/**
 * Revert command that mimics sqitch revert behavior
 * This is designed to be a drop-in replacement for spawn('sqitch', ['revert', 'db:pg:database'])
 */
export async function revertModule(
  config: Partial<MigrateConfig>,
  database: string,
  cwd: string,
  options?: {
    toChange?: string;
    useTransaction?: boolean;
  }
): Promise<void> {
  const planPath = join(cwd, 'launchql.plan');
  
  if (!existsSync(planPath)) {
    throw new Error(`No launchql.plan found in ${cwd}`);
  }
  
  // Provide defaults for missing config values
  const fullConfig: MigrateConfig = {
    host: config.host,
    port: config.port,
    user: config.user,
    password: config.password,
    database: config.database
  };
  
  const client = new LaunchQLMigrate(fullConfig);
  
  try {
    const result = await client.revert({
      project: '', // Will be read from plan file
      targetDatabase: database,
      planPath,
      toChange: options?.toChange,
      useTransaction: options?.useTransaction
    });
    
    if (result.failed) {
      throw new Error(`Revert failed at change: ${result.failed}`);
    }
    
    log.info(`Reverted ${result.reverted.length} changes`);
    if (result.skipped.length > 0) {
      log.info(`Skipped ${result.skipped.length} not deployed changes`);
    }
  } finally {
    // Pool is managed by PgPoolCacheManager, no need to close
  }
}