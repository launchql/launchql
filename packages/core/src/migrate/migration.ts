import { LaunchQLProject } from '../class/launchql';
import { deployProject } from '../projects/deploy-project';
import { revertProject } from '../projects/revert-project';
import { verifyProject } from '../projects/verify-project';
import { getEnvOptions } from '@launchql/types';
import { getPgEnvOptions } from 'pg-env';
import { Logger } from '@launchql/logger';
import { deployModule } from './deploy-module';
import { revertModule } from './revert-module';
import { verifyModule } from './verify-module';
import { runSqitch } from '../utils/sqitch-wrapper';

const log = new Logger('project-commands');

export interface MigrationOptions {
  database: string;
  cwd: string;
  recursive?: boolean;
  projectName?: string;  // Required if recursive=true
  useSqitch?: boolean;
  useTransaction?: boolean;
  toChange?: string;
  // Options for fast deployment
  fast?: boolean;
  usePlan?: boolean;
  cache?: boolean;
  /**
   * The plan file to use for sqitch operations
   * Defaults to 'launchql.plan'
   */
  planFile?: string;
}

/**
 * Deploy a project - handles both recursive (multi-module) and non-recursive (single directory) deployments
 */
export async function deployModules(options: MigrationOptions): Promise<void> {
  if (options.recursive) {
    if (!options.projectName) {
      throw new Error('projectName is required when recursive is true');
    }

    // Use existing deploy() from core that handles dependencies
    const project = new LaunchQLProject(options.cwd);
    const modules = project.getModuleMap();
    
    if (!modules[options.projectName]) {
      throw new Error(`Module "${options.projectName}" not found`);
    }
    
    const modulePath = modules[options.projectName].path;
    log.info(`Deploying project ${options.projectName} from ${modulePath} to database ${options.database}...`);
    
    await deployProject(
      getEnvOptions({ pg: { database: options.database } }), 
      options.projectName, 
      options.database, 
      modulePath, 
      { 
        useSqitch: options.useSqitch, 
        useTransaction: options.useTransaction,
        fast: options.fast,
        usePlan: options.usePlan,
        cache: options.cache,
        planFile: options.planFile,
        toChange: options.toChange
      }
    );
  } else {
    // Direct execution on current directory
    if (options.useSqitch) {
      await runSqitch('deploy', options.database, options.cwd, getPgEnvOptions(), {
        planFile: options.planFile,
        args: options.toChange ? [options.toChange] : []
      });
    } else {
      await deployModule(
        getPgEnvOptions(), 
        options.database, 
        options.cwd, 
        { 
          useTransaction: options.useTransaction, 
          toChange: options.toChange 
        }
      );
    }
  }
}

/**
 * Revert a project - handles both recursive (multi-module) and non-recursive (single directory) reverts
 */
export async function revertModules(options: MigrationOptions): Promise<void> {
  if (options.recursive) {
    if (!options.projectName) {
      throw new Error('projectName is required when recursive is true');
    }

    // Use existing revert() from core
    const project = new LaunchQLProject(options.cwd);
    const modules = project.getModuleMap();
    
    if (!modules[options.projectName]) {
      throw new Error(`Module "${options.projectName}" not found`);
    }
    
    log.info(`Reverting project ${options.projectName} on database ${options.database}...`);
    
    await revertProject(
      getEnvOptions({ pg: { database: options.database } }), 
      options.projectName, 
      options.database, 
      options.cwd, 
      { 
        useSqitch: options.useSqitch, 
        useTransaction: options.useTransaction,
        planFile: options.planFile,
        toChange: options.toChange
      }
    );
  } else {
    // Direct execution on current directory
    if (options.useSqitch) {
      await runSqitch('revert', options.database, options.cwd, getPgEnvOptions(), {
        planFile: options.planFile,
        confirm: true,
        args: options.toChange ? [options.toChange] : []
      });
    } else {
      await revertModule(
        getPgEnvOptions(), 
        options.database, 
        options.cwd, 
        { 
          useTransaction: options.useTransaction, 
          toChange: options.toChange 
        }
      );
    }
  }
}

/**
 * Verify a project - handles both recursive (multi-module) and non-recursive (single directory) verification
 */
export async function verifyModules(options: MigrationOptions): Promise<void> {
  if (options.recursive) {
    if (!options.projectName) {
      throw new Error('projectName is required when recursive is true');
    }

    // Use existing verify() from core
    const project = new LaunchQLProject(options.cwd);
    const modules = project.getModuleMap();
    
    if (!modules[options.projectName]) {
      throw new Error(`Module "${options.projectName}" not found`);
    }
    
    log.info(`Verifying project ${options.projectName} on database ${options.database}...`);
    
    await verifyProject(
      getEnvOptions({ pg: { database: options.database } }), 
      options.projectName, 
      options.database, 
      options.cwd, 
      { 
        useSqitch: options.useSqitch,
        planFile: options.planFile
      }
    );
  } else {
    // Direct execution on current directory
    if (options.useSqitch) {
      await runSqitch('verify', options.database, options.cwd, getPgEnvOptions(), {
        planFile: options.planFile
      });
    } else {
      await verifyModule(
        getPgEnvOptions(), 
        options.database, 
        options.cwd
      );
    }
  }
}

/**
 * Get available modules in a directory
 */
export async function getAvailableModules(cwd: string): Promise<string[]> {
  const project = new LaunchQLProject(cwd);
  const modules = await project.getModules();
  return modules.map(mod => mod.getModuleName());
}