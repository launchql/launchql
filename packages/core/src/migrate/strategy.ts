import { PgConfig } from 'pg-env';
import { LaunchQLOptions } from '@launchql/types';
import { deployModule } from './deploy-module';
import { revertModule } from './revert-module';
import { verifyModule } from './verify-module';
import { runSqitch } from '../utils/sqitch-wrapper';
import { packageModule } from '../package';
import { LaunchQLProject } from '../class/launchql';
import { getPgPool } from 'pg-cache';
import { Logger } from '@launchql/logger';
import { errors } from '@launchql/types';

const log = new Logger('migration-strategy');

export interface StrategyOptions {
  useSqitch?: boolean;
  useTransaction?: boolean;
  fast?: boolean;
  usePlan?: boolean;
  cache?: boolean;
  planFile?: string;
  toChange?: string;
}

const deployFastCache: Record<string, Awaited<ReturnType<typeof packageModule>>> = {};

const getCacheKey = (
  pg: PgConfig,
  name: string,
  database: string
): string => {
  const { host, port, user } = pg ?? {};
  return `${host}:${port}:${user}:${database}:${name}`;
};

async function executeFastDeploy(
  opts: LaunchQLOptions,
  database: string,
  modulePath: string,
  options: StrategyOptions,
  extensionName?: string
): Promise<void> {
  const pgPool = getPgPool({ ...opts.pg, database });
  
  const cacheKey = extensionName ? getCacheKey(opts.pg as PgConfig, extensionName, database) : '';
  
  if (options.cache && cacheKey && deployFastCache[cacheKey]) {
    log.warn(`⚡ Using cached pkg for ${extensionName}.`);
    await pgPool.query(deployFastCache[cacheKey].sql);
    return;
  }

  let pkg;
  try {
    pkg = await packageModule(modulePath, { 
      usePlan: options.usePlan ?? true, 
      extension: false 
    });
  } catch (err) {
    log.error(`❌ Failed to package module at path: ${modulePath}`);
    log.error(`   Error: ${err instanceof Error ? err.message : String(err)}`);
    throw err;
  }

  log.debug(`→ Command: sqitch deploy db:pg:${database}`);
  log.debug(`> ${pkg.sql}`);

  await pgPool.query(pkg.sql);

  if (options.cache && cacheKey) {
    deployFastCache[cacheKey] = pkg;
  }
}

export async function executeDeployStrategy(
  opts: LaunchQLOptions | PgConfig,
  database: string,
  modulePath: string,
  options: StrategyOptions = {},
  extensionName?: string
): Promise<void> {
  if (options.fast ?? true) {
    await executeFastDeploy(opts as LaunchQLOptions, database, modulePath, options, extensionName);
  } else if (options.useSqitch) {
    const planFile = options.planFile || 'launchql.plan';
    const sqitchArgs = options.toChange ? [options.toChange] : [];
    log.debug(`→ Command: sqitch deploy --plan-file ${planFile} db:pg:${database}${sqitchArgs.length ? ' ' + sqitchArgs.join(' ') : ''}`);
    
    try {
      const exitCode = await runSqitch('deploy', database, modulePath, opts as PgConfig, {
        planFile,
        args: sqitchArgs
      });
      
      if (exitCode !== 0) {
        log.error(`❌ Deployment failed for module ${extensionName || 'unknown'}`);
        throw errors.DEPLOYMENT_FAILED({ type: 'Deployment', module: extensionName || 'unknown' });
      }
    } catch (err) {
      log.error(`❌ Deployment failed for module ${extensionName || 'unknown'}`);
      throw err;
    }
  } else {
    log.debug(`→ Command: launchql migrate deploy db:pg:${database}`);
    
    try {
      await deployModule(opts as PgConfig, database, modulePath, { 
        useTransaction: options.useTransaction,
        toChange: options.toChange
      });
    } catch (deployError) {
      log.error(`❌ Deployment failed for module ${extensionName || 'unknown'}`);
      throw deployError;
    }
  }
}

export async function executeRevertStrategy(
  opts: LaunchQLOptions | PgConfig,
  database: string,
  modulePath: string,
  options: StrategyOptions = {},
  extensionName?: string
): Promise<void> {
  if (options.useSqitch) {
    const planFile = options.planFile || 'launchql.plan';
    const sqitchArgs = options.toChange ? [options.toChange] : [];
    log.debug(`→ Command: sqitch revert --plan-file ${planFile} db:pg:${database} -y${sqitchArgs.length ? ' ' + sqitchArgs.join(' ') : ''}`);

    try {
      const exitCode = await runSqitch('revert', database, modulePath, opts as PgConfig, {
        planFile,
        confirm: true,
        args: sqitchArgs
      });
      
      if (exitCode !== 0) {
        log.error(`❌ Revert failed for module ${extensionName || 'unknown'}`);
        throw errors.DEPLOYMENT_FAILED({ type: 'Revert', module: extensionName || 'unknown' });
      }
    } catch (err) {
      log.error(`❌ Revert failed for module ${extensionName || 'unknown'}`);
      throw err;
    }
  } else {
    log.debug(`→ Command: launchql migrate revert db:pg:${database}`);
    
    try {
      await revertModule(opts as PgConfig, database, modulePath, { 
        useTransaction: options.useTransaction,
        toChange: options.toChange
      });
    } catch (revertError) {
      log.error(`❌ Revert failed for module ${extensionName || 'unknown'}`);
      throw revertError;
    }
  }
}

export async function executeVerifyStrategy(
  opts: LaunchQLOptions | PgConfig,
  database: string,
  modulePath: string,
  options: StrategyOptions = {},
  extensionName?: string
): Promise<void> {
  if (options.useSqitch) {
    const planFile = options.planFile || 'launchql.plan';
    log.debug(`→ Command: sqitch verify --plan-file ${planFile} db:pg:${database}`);
    
    try {
      const exitCode = await runSqitch('verify', database, modulePath, opts as PgConfig, {
        planFile
      });
      
      if (exitCode !== 0) {
        throw new Error(`sqitch verify exited with code ${exitCode}`);
      }
    } catch (err) {
      throw new Error(`Verification failed: ${err instanceof Error ? err.message : String(err)}`);
    }
  } else {
    await verifyModule(opts as PgConfig, database, modulePath);
  }
}
