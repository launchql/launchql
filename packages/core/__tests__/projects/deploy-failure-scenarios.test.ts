import { LaunchQLMigrate } from '../../src/migrate/client';
import { MigrateTestFixture, teardownAllPools, TestDatabase } from '../../test-utils';

describe('Deploy Failure Scenarios', () => {
  let fixture: MigrateTestFixture;
  let db: TestDatabase;
  
  beforeEach(async () => {
    fixture = new MigrateTestFixture();
    db = await fixture.setupTestDatabase();
  });
  
  afterEach(async () => {
    await fixture.cleanup();
  });

  afterAll(async () => {
    await teardownAllPools();
  });

  
  test('constraint violation with transaction - automatic rollback', async () => {
    /*
     * SCENARIO: Transaction-based deployment with constraint violation
     * 
     * This test demonstrates LaunchQL's automatic rollback behavior when useTransaction: true (default).
     * When ANY change fails during deployment, ALL changes are automatically rolled back.
     * 
     * Expected behavior:
     * - All 3 changes attempted in single transaction
     * - Constraint violation on 3rd change triggers complete rollback
     * - Database state: clean (as if deployment never happened)
     * - Migration tracking: zero deployed changes, failure events logged
     */
    const tempDir = fixture.createPlanFile('test-constraint-fail', [
      { name: 'create_table' },
      { name: 'add_constraint', dependencies: ['create_table'] },
      { name: 'violate_constraint', dependencies: ['add_constraint'] }
    ]);
    
    fixture.createScript(tempDir, 'deploy', 'create_table', 
      'CREATE TABLE test_users (id SERIAL PRIMARY KEY, email VARCHAR(255) UNIQUE);'
    );
    
    fixture.createScript(tempDir, 'deploy', 'add_constraint', 
      "INSERT INTO test_users (email) VALUES ('test@example.com');"
    );
    
    fixture.createScript(tempDir, 'deploy', 'violate_constraint', 
      "INSERT INTO test_users (email) VALUES ('test@example.com');"
    );
    
    const client = new LaunchQLMigrate(db.config);
    
    const initialState = await db.getMigrationState();
    expect(initialState.changeCount).toBe(0);
    expect(initialState.eventCount).toBe(0);
    
    await expect(client.deploy({
      modulePath: tempDir,
      useTransaction: true
    })).rejects.toThrow(/duplicate key value violates unique constraint/);
    
    const finalState = await db.getMigrationState();
    
    expect(finalState).toMatchSnapshot('transaction-rollback-migration-state');
    
    expect(finalState.changeCount).toBe(0);
    expect(finalState.eventCount).toBe(1); // Now expect deploy failure event to be logged
    expect(finalState.events[0].event_type).toBe('deploy');
    expect(finalState.events[0].error_message).toContain('duplicate key value violates unique constraint');
    
    expect(await db.exists('table', 'test_users')).toBe(false);
  });
  
  test('constraint violation without transaction - partial deployment', async () => {
    /*
     * SCENARIO: Non-transaction deployment with constraint violation
     * 
     * This test demonstrates LaunchQL's behavior when useTransaction: false.
     * Each change is deployed individually - successful changes remain deployed
     * even when later changes fail. Manual cleanup is required.
     * 
     * Expected behavior:
     * - Changes deployed one-by-one (no transaction wrapper)
     * - First 2 changes succeed and remain deployed
     * - 3rd change fails on constraint violation, deployment stops
     * - 4th change never attempted (deployment stops at first failure)
     * - Database state: partial (successful changes persist)
     * - Migration tracking: shows successful deployments + failure events
     */
    const tempDir = fixture.createPlanFile('test-constraint-partial', [
      { name: 'create_table' },
      { name: 'add_record', dependencies: ['create_table'] },
      { name: 'violate_constraint', dependencies: ['add_record'] },
      { name: 'final_change', dependencies: ['add_record'] }
    ]);
    
    fixture.createScript(tempDir, 'deploy', 'create_table', 
      'CREATE TABLE test_products (id SERIAL PRIMARY KEY, sku VARCHAR(50) UNIQUE);'
    );
    
    fixture.createScript(tempDir, 'deploy', 'add_record', 
      "INSERT INTO test_products (sku) VALUES ('PROD-001');"
    );
    
    fixture.createScript(tempDir, 'deploy', 'violate_constraint', 
      "INSERT INTO test_products (sku) VALUES ('PROD-001');"
    );
    
    fixture.createScript(tempDir, 'deploy', 'final_change', 
      "INSERT INTO test_products (sku) VALUES ('PROD-002');"
    );
    
    const client = new LaunchQLMigrate(db.config);
    
    const initialState = await db.getMigrationState();
    expect(initialState.changeCount).toBe(0);
    
    await expect(client.deploy({
      modulePath: tempDir,
      useTransaction: false
    })).rejects.toThrow(/duplicate key value violates unique constraint/);
    
    const finalState = await db.getMigrationState();
    
    expect(finalState).toMatchSnapshot('partial-deployment-migration-state');
    
    expect(finalState.changeCount).toBe(2);
    expect(finalState.changes.map((c: any) => c.change_name)).toEqual(['create_table', 'add_record']);
    
    expect(await db.exists('table', 'test_products')).toBe(true);
    const records = await db.query('SELECT * FROM test_products');
    expect(records.rows).toHaveLength(1);
    expect(records.rows[0].sku).toBe('PROD-001');
    
    const finalRecord = await db.query("SELECT * FROM test_products WHERE sku = 'PROD-002'");
    expect(finalRecord.rows).toHaveLength(0);
    
    const successEvents = finalState.events.filter((e: any) => e.event_type === 'deploy' && !e.error_message);
    expect(successEvents.length).toBe(2); // create_table, add_record
    const failEvents = finalState.events.filter((e: any) => e.event_type === 'deploy' && e.error_message);
    expect(failEvents.length).toBe(1); // violate_constraint failure logged
    expect(finalState.eventCount).toBe(3); // 2 successful deployments + 1 failure
  });
  
  test('transaction mode - complete rollback on constraint failure', async () => {
    /*
     * SCENARIO: Transaction-based deployment with constraint failure
     * 
     * This test demonstrates LaunchQL's complete rollback behavior when useTransaction: true (default).
     * When ANY change fails during deployment, ALL changes are automatically rolled back.
     * 
     * Expected behavior:
     * - All 3 changes attempted in single transaction
     * - Constraint violation on 3rd change triggers complete rollback
     * - Database state: clean (as if deployment never happened)
     * - Migration tracking: zero deployed changes, failure event logged outside transaction
     */
    const tempDir = fixture.createPlanFile('test-transaction-rollback', [
      { name: 'setup_schema' },
      { name: 'create_constraint_table', dependencies: ['setup_schema'] },
      { name: 'fail_on_constraint', dependencies: ['create_constraint_table'] }
    ]);
    
    fixture.createScript(tempDir, 'deploy', 'setup_schema', 
      'CREATE SCHEMA test_schema;'
    );
    
    fixture.createScript(tempDir, 'deploy', 'create_constraint_table', 
      'CREATE TABLE test_schema.orders (id SERIAL PRIMARY KEY, amount DECIMAL(10,2) CHECK (amount > 0));'
    );
    
    fixture.createScript(tempDir, 'deploy', 'fail_on_constraint', 
      'INSERT INTO test_schema.orders (amount) VALUES (-100.00);'
    );
    
    const client = new LaunchQLMigrate(db.config);
    
    await expect(client.deploy({
      modulePath: tempDir,
      useTransaction: true
    })).rejects.toThrow(/violates check constraint/);
    
    const finalState = await db.getMigrationState();
    
    expect(finalState).toMatchSnapshot('transaction-mode-constraint-failure');
    
    expect(await db.exists('schema', 'test_schema')).toBe(false);
    expect(finalState.changeCount).toBe(0);
    expect(finalState.eventCount).toBe(1); // Deploy failure event logged outside transaction
    expect(finalState.events[0].event_type).toBe('deploy');
    expect(finalState.events[0].error_message).toContain('violates check constraint');
    
    /*
     * KEY INSIGHT: Transaction mode provides complete rollback
     * 
     * - launchql_migrate.changes: 0 rows (complete rollback)
     * - launchql_migrate.events: 1 failure event (logged outside transaction)
     * - Database objects: none (clean state)
     * 
     * RECOMMENDATION: Use transaction mode (default) for atomic deployments
     * where you want all-or-nothing behavior.
     */
  });

  test('non-transaction mode - partial deployment on constraint failure', async () => {
    /*
     * SCENARIO: Non-transaction deployment with constraint failure
     * 
     * This test demonstrates LaunchQL's partial deployment behavior when useTransaction: false.
     * Each change is deployed individually - successful changes remain deployed
     * even when later changes fail. Deployment stops at first failure.
     * 
     * Expected behavior:
     * - Changes deployed one-by-one (no transaction wrapper)
     * - First 2 changes succeed and remain deployed
     * - 3rd change fails on constraint violation, deployment stops
     * - Database state: partial (successful changes persist)
     * - Migration tracking: shows successful deployments + failure event
     */
    const tempDir = fixture.createPlanFile('test-nontransaction-partial', [
      { name: 'setup_schema' },
      { name: 'create_constraint_table', dependencies: ['setup_schema'] },
      { name: 'fail_on_constraint', dependencies: ['create_constraint_table'] }
    ]);
    
    fixture.createScript(tempDir, 'deploy', 'setup_schema', 
      'CREATE SCHEMA test_schema;'
    );
    
    fixture.createScript(tempDir, 'deploy', 'create_constraint_table', 
      'CREATE TABLE test_schema.orders (id SERIAL PRIMARY KEY, amount DECIMAL(10,2) CHECK (amount > 0));'
    );
    
    fixture.createScript(tempDir, 'deploy', 'fail_on_constraint', 
      'INSERT INTO test_schema.orders (amount) VALUES (-100.00);'
    );
    
    const client = new LaunchQLMigrate(db.config);
    
    await expect(client.deploy({
      modulePath: tempDir,
      useTransaction: false
    })).rejects.toThrow(/violates check constraint/);
    
    const finalState = await db.getMigrationState();
    
    expect(finalState).toMatchSnapshot('non-transaction-mode-constraint-failure');
    
    expect(await db.exists('schema', 'test_schema')).toBe(true);
    expect(await db.exists('table', 'test_schema.orders')).toBe(true);
    
    const records = await db.query('SELECT * FROM test_schema.orders');
    expect(records.rows).toHaveLength(0);
    
    expect(finalState.changeCount).toBe(2);
    expect(finalState.changes.map((c: any) => c.change_name)).toEqual(['setup_schema', 'create_constraint_table']);
    
    const successEvents = finalState.events.filter((e: any) => e.event_type === 'deploy' && !e.error_message);
    expect(successEvents.length).toBe(2); // setup_schema, create_constraint_table
    const failEvents = finalState.events.filter((e: any) => e.event_type === 'deploy' && e.error_message);
    expect(failEvents.length).toBe(1); // fail_on_constraint failure logged
    expect(finalState.eventCount).toBe(3); // 2 successful deployments + 1 failure
    
    /*
     * KEY INSIGHT: Non-transaction mode provides partial deployment
     * 
     * - launchql_migrate.changes: 2 rows (partial success)
     * - launchql_migrate.events: 2 success + 1 failure event
     * - Database objects: schema + table exist (mixed state)
     * 
     * IMPORTANT: Deployment stops immediately at first failure, just like transaction mode.
     * The difference is in state persistence, not error handling behavior.
     * 
     * RECOMMENDATION: Use non-transaction mode only when you specifically
     * need partial deployment behavior for incremental rollout scenarios.
     */
  });

  test('verify failure - non-existent table reference', async () => {
    /*
     * SCENARIO: Verify script references a table that doesn't exist
     * 
     * This test demonstrates LaunchQL's verify failure behavior when a verify script
     * tries to check a table that was never created or was dropped.
     * 
     * Expected behavior:
     * - Deploy succeeds (creates a simple table)
     * - Verify fails because script references non-existent table
     * - Failure event logged with detailed error information
     */
    const tempDir = fixture.createPlanFile('test-verify-fail', [
      { name: 'create_simple_table' }
    ]);
    
    fixture.createScript(tempDir, 'deploy', 'create_simple_table', 
      'CREATE TABLE users (id SERIAL PRIMARY KEY, name VARCHAR(100));'
    );
    
    fixture.createScript(tempDir, 'verify', 'create_simple_table', 
      'SELECT 1 FROM non_existent_table LIMIT 1;'
    );
    
    const client = new LaunchQLMigrate(db.config);
    
    await client.deploy({
      modulePath: tempDir,
      useTransaction: true
    });
    
    const deployState = await db.getMigrationState();
    expect(deployState.changeCount).toBe(1);
    expect(await db.exists('table', 'users')).toBe(true);
    
    const verifyResult = await client.verify({
      modulePath: tempDir
    });
    
    expect(verifyResult.verified).toEqual([]);
    expect(verifyResult.failed).toEqual(['create_simple_table']);
    
    const finalState = await db.getMigrationState();
    
    expect(finalState).toMatchSnapshot('verify-failure-non-existent-table');
    
    // Should have deploy success event + verify failure event
    const verifyEvents = finalState.events.filter((e: any) => e.event_type === 'verify');
    expect(verifyEvents.length).toBe(1);
    expect(verifyEvents[0].error_message).toBe('Verification failed');
    expect(verifyEvents[0].error_code).toBe(null);
  });
});
