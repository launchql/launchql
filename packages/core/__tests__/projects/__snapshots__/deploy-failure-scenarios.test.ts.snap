// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Deploy Failure Scenarios constraint violation with transaction - automatic rollback: transaction-rollback-migration-state 1`] = `
{
  "changeCount": 0,
  "changes": [],
  "eventCount": 0,
  "events": [],
}
`;

exports[`Deploy Failure Scenarios constraint violation without transaction - partial deployment: partial-deployment-migration-state 1`] = `
{
  "changeCount": 2,
  "changes": [
    {
      "change_name": "add_record",
      "project": "test-constraint-partial",
      "script_hash": "833d7d349e3c4f07e1a24ed40ac9814329efc87c180180342a09874f8124a037",
    },
    {
      "change_name": "create_table",
      "project": "test-constraint-partial",
      "script_hash": "0624b3e2276299c8c3b8bfa514fe0d128906193769b3aeaea6732e71c0e352e6",
    },
  ],
  "eventCount": 2,
  "events": [
    {
      "change_name": "add_record",
      "event_type": "deploy",
      "project": "test-constraint-partial",
    },
    {
      "change_name": "create_table",
      "event_type": "deploy",
      "project": "test-constraint-partial",
    },
  ],
}
`;

exports[`Deploy Failure Scenarios verify database state after constraint failure: partial-deployment-state-comparison 1`] = `
{
  "changeCount": 2,
  "changes": [
    {
      "change_name": "create_constraint_table",
      "project": "test-state-check",
      "script_hash": "ec5b17e155a2cd4e098716204192083d31d096a4cf163550d5ea176a4615a4d2",
    },
    {
      "change_name": "setup_schema",
      "project": "test-state-check",
      "script_hash": "a3419a48994fd13a668befcaab23c4d0d7e9e08e6e6a9093effb3c85b7e953d9",
    },
  ],
  "eventCount": 2,
  "events": [
    {
      "change_name": "create_constraint_table",
      "event_type": "deploy",
      "project": "test-state-check",
    },
    {
      "change_name": "setup_schema",
      "event_type": "deploy",
      "project": "test-state-check",
    },
  ],
}
`;

exports[`Deploy Failure Scenarios verify database state after constraint failure: transaction-rollback-state-comparison 1`] = `
{
  "changeCount": 0,
  "changes": [],
  "eventCount": 0,
  "events": [],
}
`;
